### DISCUSSION TITLE
Load Testing, CPU Cores And Sample ULCL Config

### DISCUSSION THREAD

**Discussion by infinitydon** [edgecomllc/eupf#487] [Q&A] [answered]
Hi,

I just came across your UPF implementation and will like to clarify the following:

- Why is the throughput test lower than Open5gs and Free5gc UPFs? I thought been an eBPF based UPF should have much more performance benefits
- Does the eUPF make use of multiple CPU cores to increase the performance?
- Are there any reference ULCL sample configuration?

Thanks

**Comment by pirog-spb** [comment]
Hi @infinitydon,

> Why is the throughput test lower than Open5gs and Free5gc UPFs? I thought been an eBPF based UPF should have much more performance benefits

The test results a little bit outdated, actually. But the main reason of poor perfomance was using eBPF XDP program in generic attach mode. In this mode no eBPF benefits are avaliable. In pure K8s deployment with veth interfaces generic mode is the only option.

In order to achieve better performance results is necessary to use at least native attach mode. We are going to share new performance tests results on bare metal environment where native mode was used. 

> Does the eUPF make use of multiple CPU cores to increase the performance?

Yes. It corresponds to rx queues count configured in network interface(port).

> Are there any reference ULCL sample configuration?

Please, check deployments examples [in the repository](https://github.com/edgecomllc/eupf/tree/main/docs/deployments)
 
--BR Alex

**Reply by infinitydon** [reply]
Thanks Alex for the response..

As for the k8s aspect, I think using multus with the [host-dev plugin](https://www.cni.dev/plugins/current/main/host-device/) i.e. dedicating a full NIC should be the proper way to get full performance.

Also with CNI plugins that are based on eBPF like Calico and Cilium, how do you think they will be of maximum benefit with respect to using for eUPF in k8s?

**Comment by pirog-spb** [comment, marked as answer]
Of course, using PCI-Passthrough or SR-IOV will definitely improve performance, but cloud compatibility/nativeness is less in this case.

I think, using eBPF native datapath from Cilium/Calico could improve performace as well. For this option it's possible to use XDP eBPF in **native mode** for veth driver in container(eBPF XDP programs have to be loaded on both interfaces of veth pair to use native mode).

![image](https://github.com/edgecomllc/eupf/assets/20152142/bdbd8925-961e-4231-a815-44f599c49d5e)

--BR Alex

**Comment by pirog-spb** [comment]
BTW, I'll add the issue to support mentioned scenario in our backlog

**Reply by infinitydon** [reply]
Ok.. Thanks
